components/mediation-monitor/mediation-data-publisher/org.wso2.carbon.prometheus.publisher/src/main/java/org/wso2/carbon/prometheus/publisher/publisher/MetricCollector.java

```diff
@@ -0,0 +1,350 @@
+
+    private static final String SYNAPSE_DOMAIN_NAME = "org.apache.synapse:*";
+    private static final String METRIC_TYPE_UNTYPED = "untyped";
+    private static final String SEPARATOR = "_";
+    private static final Pattern PROPERTY_PATTERN = Pattern.compile(
+            "([^,=:*?]+)" + // Name - non-empty, anything but comma, equals, colon, star, or question mark
+                    "=" +  // Equals
+                    "(" + // Either
+                    "\"" + // Quoted
+                    "(?:" + // A possibly empty sequence of
+                    "[^\\\\\"]*" + // Greedily match anything but backslash or quote
+                    "(?:\\\\.)?" + // Greedily see if we can match an escaped sequence
+                    ")*" +
+                    "\"" +
+                    "|" + // Or
+                    "[^,=:\"]*" + // Unquoted - can be empty, anything but comma, equals, colon, or quote
+                    ")");
+
+    private static volatile MetricCollector metricCollector;
+    private List<PrometheusMetric> metricList;
+    private Map<ObjectName, LinkedHashMap<String, String>> keyPropertiesPerBean;
+
+    private MetricCollector() {
+
+        this.metricList = new ArrayList<>();
+        this.keyPropertiesPerBean = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * @return Single instance of MetricCollector
+     */
+    static MetricCollector getInstance() {
+
+        MetricCollector collector = metricCollector;
+        if (collector == null) {
+            synchronized (MetricCollector.class) {
+                collector = metricCollector;
+                if (collector == null) {
+                    metricCollector = collector = new MetricCollector();
+                }
+            }
+        }
+        return collector;
+    }
+
+    /**
+     * Collects metric data from JMX MBeans
+     *
+     * @return Metric data in Prometheus format
+     */
+    List<String> collect() {
+
+        this.metricList = new ArrayList<>();
+        this.keyPropertiesPerBean = new ConcurrentHashMap<>();
+
+        try {
+
+            MBeanServer beanServer = ManagementFactory.getPlatformMBeanServer();
+
+            Set<ObjectName> mBeanNames = new HashSet<>();
+
+            // get filtered MBean names for Synapse domain
+            ObjectName synapseFilterName = new ObjectName(SYNAPSE_DOMAIN_NAME);
+            for (ObjectInstance instance : beanServer.queryMBeans(synapseFilterName, null)) {
+                mBeanNames.add(instance.getObjectName());
+            }
+
+            // scrape each MBean
+            for (ObjectName beanName : mBeanNames) {
+                scrapeBean(beanServer, beanName);
+            }
+
+            // sort metrics so that related metrics are grouped together
+            if (this.metricList != null && !metricList.isEmpty()) {
+                Collections.sort(metricList, new Comparator<PrometheusMetric>() {
```

Comment:
Shall we use lambda syntax here?

Reply:
Lambda syntax cannot be used as the Apache maven bundle plugin supports only Java 7. Bundle creation fails when Java 8 is used.

Reply:
We are using JDK 8 in the maven-compiler-plugin. Therefore I believe we should be able to use JDK 8 features.

[1]. https://github.com/wso2/carbon-mediation/blob/277003b0f5c14b1fcb251ddf55ac6d03da88461a/components/mediation-monitor/mediation-data-publisher/org.wso2.carbon.prometheus.publisher/pom.xml#L66